### PR TITLE
fix #4943

### LINKED ISSUE #4943: Visual block replace operation over non-ASCII characters crashes Neovim
- `nvim --version`: 

```
NVIM v0.1.5-480-gd5ddebe
Build type: Dev
Compilation: /usr/bin/cc -march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong --param=ssp-buffer-size=4 -Wconversion -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wvla -fstack-protector-strong -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -DHAVE_CONFIG_H -D_GNU_SOURCE -I/build/neovim-git/src/neovim-git/build/config -I/build/neovim-git/src/neovim-git/src -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/usr/include -I/build/neovim-git/src/neovim-git/build/src/nvim/auto -I/build/neovim-git/src/neovim-git/build/include
Compiled by zoqaeski@computer
    
Optional features included (+) or not (-): +acl   +iconv    +jemalloc +tui
For differences from Vim, see :help vim-differences
    
system vimrc file: "$VIM/sysinit.vim"
fall-back for $VIM: "/usr/share/nvim"
````    

- Vim (version: ) behaves differently? Yes
- Operating system/version: Arch Linux x86-64
- Terminal name/version: Terminix 1.0.0 (also occurs with neovim-qt-git r332.adc7a2d)
- `$TERM`: screen-256color

### Actual behaviour
Typing a malformed Compose key sequence causes neovim to crash.

### Expected behaviour
Neovim drops invalid key sequence and silently ignores it.

### Steps to reproduce using `nvim -u NORC`

```
nvim -u NORC
i
hahača
hahača
hahača
hahača
hahača
```
Visual block select the `č`s and try to replace it with an invalid sequence (Compose m m is one), that is `r Compose m m`. You get a segfault op_replace 'n >= 0' failed in ops.c:1689, and the terminal is completely borked, spewing out loads of escape sequences.

I tested it a number of times: 
* just doing a straight replace from normal mode 
* replace from visual mode 
* replace from visual block mode
* plain insert mode

The crash only occurs in visual block mode and only when you try to replace a non-ASCII character.

**EDIT** It looks like this is occurring regardless of what you replace the non-ASCII character with.
```
nvim -u NORC
5iač<CR><ESC>ggl<C-v>4jrm
```
CRASH but
```
nvim -u NORC
5iac<CR><ESC>ggl<C-v>4jrm
```
works.

### PR DESCRIPTION
**Pull request by cjoach**
Fix #4943 by changing the cast appropriately.

### CODE REVIEW

File: src/nvim/ops.c [hunk 1]
```diff
@@ -1686,9 +1686,9 @@ int op_replace(oparg_T *oap, int c)
 
       oldp = get_cursor_line_ptr();
       oldlen = STRLEN(oldp);
-      assert(n >= 0);
-      newp = (char_u *)xmalloc(oldlen + 1 + (size_t)n);
-      memset(newp, NUL, oldlen + 1 + (size_t)n);
+      assert(oldlen <= INT_MAX);
```
Comment (justinmk) [inline src/nvim/ops.c]: Should this be:

    assert(oldlen <= INT_MAX - 1 - n);

Comment (cacplate) [inline src/nvim/ops.c]: indeed.

File: src/nvim/ops.c [hunk 2]
```diff
@@ -1686,9 +1686,9 @@ int op_replace(oparg_T *oap, int c)
 
       oldp = get_cursor_line_ptr();
       oldlen = STRLEN(oldp);
-      assert(n >= 0);
-      newp = (char_u *)xmalloc(oldlen + 1 + (size_t)n);
-      memset(newp, NUL, oldlen + 1 + (size_t)n);
+      assert(oldlen <= INT_MAX - 1 - n);
+      newp = (char_u *)xmalloc((size_t)((int)oldlen + 1 + n));
```
Comment (justinmk) [inline src/nvim/ops.c]: Can the cast `(char_u *)` be removed?

File: src/nvim/ops.c [hunk 3]
```diff
@@ -1686,9 +1686,9 @@ int op_replace(oparg_T *oap, int c)
 
       oldp = get_cursor_line_ptr();
       oldlen = STRLEN(oldp);
-      assert(n >= 0);
-      newp = (char_u *)xmalloc(oldlen + 1 + (size_t)n);
-      memset(newp, NUL, oldlen + 1 + (size_t)n);
+      assert(oldlen <= INT_MAX - 1 - n);
```
Comment (justinmk) [inline src/nvim/ops.c]: Looking at it some more, perhaps:

```c
  int new_size = (int)oldlen + 1 + n
  assert(new_size >= 0 && oldlen <= INT_MAX - 1 - n);
  newp = xmalloc((size_t)new_size);
  memset(newp, NUL, (size_t)new_size);
```

File: src/nvim/ops.c (hunk 2 above)
Comment (cacplate) [inline src/nvim/ops.c]: It doesn't seem to cause problem indeed.

File: src/nvim/ops.c (hunk 3 above)
Comment (cacplate) [inline src/nvim/ops.c]: Added a cast but otherwise took your idea.

File: src/nvim/ops.c [hunk 4]
```diff
@@ -1686,9 +1686,10 @@ int op_replace(oparg_T *oap, int c)
 
       oldp = get_cursor_line_ptr();
       oldlen = STRLEN(oldp);
-      assert(n >= 0);
-      newp = (char_u *)xmalloc(oldlen + 1 + (size_t)n);
-      memset(newp, NUL, oldlen + 1 + (size_t)n);
+      int new_size = (int)oldlen + 1 + n;
+      assert(new_size >= 0 && oldlen <= (size_t)(INT_MAX - 1 - n));
```
Comment (oni-link) [inline src/nvim/ops.c]: If there was an overflow in the computation of the initial value of `new_size`, the overflow check here is too late.

File: src/nvim/ops.c [hunk 5]
```diff
@@ -1686,9 +1686,10 @@ int op_replace(oparg_T *oap, int c)
...
+      newp = xmalloc((size_t)new_size);
+      memset(newp, NUL, (size_t)new_size);
```
Comment (oni-link) [inline src/nvim/ops.c]: This two lines could be replaced with `newp = xcalloc(new_size,1);`.

**Comment by cacplate** [comment]
The failing travis doesn't seem related. If there is no other comment, can we go forward with it?

File: src/nvim/ops.c [hunk 6]
```diff
@@ -1686,9 +1686,10 @@ int op_replace(oparg_T *oap, int c)
 
       oldp = get_cursor_line_ptr();
       oldlen = STRLEN(oldp);
-      assert(n >= 0);
-      newp = (char_u *)xmalloc(oldlen + 1 + (size_t)n);
-      memset(newp, NUL, oldlen + 1 + (size_t)n);
+      assert(oldlen <= (size_t)(INT_MAX - 1 - n));
```
Comment (justinmk) [inline src/nvim/ops.c]: If `n` is e.g. `-3` this check will fail  (e.g. `INT_MAX -1 - (-3)` => ). I guess we need:

    assert(n < 0 || oldlen <= (size_t)(INT_MAX - 1 - n));

_Or_ let's just drop this check and say that the assert below is enough. This is getting ridiculous :>

**Comment by justinmk** [comment]
@oni-link Any more comments?

File: src/nvim/ops.c (hunk 6 above)
Comment (oni-link) [inline src/nvim/ops.c]: I looked at the code after `xcalloc()` to see how big `newp` needs to be. Looks like one could also remove a lot of `strlen()` calls.
```diff
diff --git a/src/nvim/ops.c b/src/nvim/ops.c
index 51cb8a0..3d09142 100644
--- a/src/nvim/ops.c
+++ b/src/nvim/ops.c
@@ -1681,15 +1681,19 @@ int op_replace(oparg_T *oap, int c)
       /* Compute bytes needed, move character count to num_chars. */
       num_chars = numc;
       numc *= (*mb_char2len)(c);
-      /* oldlen includes textlen, so don't double count */
-      n += numc - bd.textlen;
 
       oldp = get_cursor_line_ptr();
       oldlen = STRLEN(oldp);
-      assert(oldlen <= (size_t)(INT_MAX - 1 - n));
-      int new_size = (int)oldlen + 1 + n;
-      assert(new_size >= 0);
-      newp = xcalloc((size_t)new_size, 1);
+
+      size_t newp_size = (size_t)(bd.textcol + bd.startspaces);
+      if (had_ctrl_v_cr || (c != '\r' && c != '\n')) {
+        newp_size += (size_t)numc;
+        if (!bd.is_short) {
+          newp_size += bd.endspaces + oldlen - bd.textcol - bd.textlen + 1;
+        }
+      }
+      newp = xcalloc(newp_size, 1);
+
       // copy up to deleted part
       memmove(newp, oldp, (size_t)bd.textcol);
       oldp += bd.textcol + bd.textlen;
@@ -1697,28 +1701,36 @@ int op_replace(oparg_T *oap, int c)
       memset(newp + bd.textcol, ' ', (size_t)bd.startspaces);
       // insert replacement chars CHECK FOR ALLOCATED SPACE
       // -1/-2 is used for entering CR literally.
+      size_t after_p_len = 0;
       if (had_ctrl_v_cr || (c != '\r' && c != '\n')) {
-        if (has_mbyte) {
-          n = (int)STRLEN(newp);
-          while (--num_chars >= 0)
-            n += (*mb_char2bytes)(c, newp + n);
-        } else
-          memset(newp + STRLEN(newp), c, (size_t)numc);
-        if (!bd.is_short) {
-          /* insert post-spaces */
-          memset(newp + STRLEN(newp), ' ', (size_t)bd.endspaces);
-          /* copy the part after the changed part */
-          STRMOVE(newp + STRLEN(newp), oldp);
+          // strlen(newp) at this point
+          int newp_len = bd.textcol + bd.startspaces;
+          if (has_mbyte) {
+            while (--num_chars >= 0) {
+              newp_len += (*mb_char2bytes)(c, newp + newp_len);
+            }
+          } else {
+            memset(newp + newp_len, c, (size_t)numc);
+            newp_len += numc;
+          }
+          if (!bd.is_short) {
+            // insert post-spaces
+            memset(newp + newp_len, ' ', (size_t)bd.endspaces);
+            newp_len += bd.endspaces;
+            // copy the part after the changed part
+            memmove(newp + newp_len, oldp,
+                    oldlen - bd.textcol - bd.textlen + 1);
         }
       } else {
         // Replacing with \r or \n means splitting the line.
-        after_p = (char_u *)xmalloc(oldlen + 1 + (size_t)n - STRLEN(newp));
-        STRMOVE(after_p, oldp);
+        after_p_len = (size_t)(oldlen - bd.textcol - bd.textlen + 1);
+        after_p = (char_u *)xmalloc(after_p_len);
+        memmove(after_p, oldp, after_p_len);
       }
       /* replace the line */
       ml_replace(curwin->w_cursor.lnum, newp, FALSE);
       if (after_p != NULL) {
-        ml_append(curwin->w_cursor.lnum++, after_p, 0, FALSE);
+        ml_append(curwin->w_cursor.lnum++, after_p, (int)after_p_len, FALSE);
         appended_lines_mark(curwin->w_cursor.lnum, 1L);
         oap->end.lnum++;
         xfree(after_p);

```

File: src/nvim/ops.c [hunk 7]
```diff
@@ -1681,43 +1681,56 @@ int op_replace(oparg_T *oap, int c)
...
-      newp = (char_u *)xmalloc(oldlen + 1 + (size_t)n);
-      memset(newp, NUL, oldlen + 1 + (size_t)n);
+      oldlen = (int)STRLEN(oldp);
+
+      size_t newp_size = (size_t)(bd.textcol + bd.startspaces);
+      if (had_ctrl_v_cr || (c != '\r' && c != '\n')) {
+        newp_size += (size_t)numc;
+        if (!bd.is_short) {
+          newp_size += (size_t)(bd.endspaces + oldlen
+                                - bd.textcol - bd.textlen + 1);
```
Comment (oni-link) [inline src/nvim/ops.c]: The `+1` here is for the terminating `NUL`. We need it always not just in this case. To add it, we could move the `+1` to the declaration of `newp_size` or remove it and use `newp = xmallocz(newp_size);`.

Problem can be seen with `5ia<C-v>u1234<cr><esc>ggl4jr<cr>` and ASAN
```c
READ of size 2 at 0x60200002d531 thread T0
    #0 0x464379 in __interceptor_strlen.part.25 /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:227
    #1 0xe483f4 in ml_flush_line /home/oni-link/git/neovim/src/nvim/memline.c:2647:26
    #2 0xe433e2 in ml_append /home/oni-link/git/neovim/src/nvim/memline.c:1836:5
    #3 0x10a6e65 in op_replace /home/oni-link/git/neovim/src/nvim/ops.c:1733:9
    #4 0xfa9924 in do_pending_operator /home/oni-link/git/neovim/src/nvim/normal.c:1965:9
    #5 0x10694f3 in normal_finish_command /home/oni-link/git/neovim/src/nvim/normal.c:890:3
    #6 0xfdfd6c in normal_execute /home/oni-link/git/neovim/src/nvim/normal.c:1147:3
    #7 0x166f3d0 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:56:26
    #8 0xf94260 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #9 0xd55611 in main /home/oni-link/git/neovim/src/nvim/main.c:541:3
    #10 0x7fb5483095af in __libc_start_main (/lib64/libc.so.6+0x205af)
    #11 0x444f08 in _start (/home/oni-link/git/neovim/build/bin/nvim+0x444f08)

0x60200002d531 is located 0 bytes to the right of 1-byte region [0x60200002d530,0x60200002d531)
allocated by thread T0 here:
    #0 0x4f66d0 in calloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:72
    #1 0xe786a2 in xcalloc /home/oni-link/git/neovim/src/nvim/memory.c:119:15
    #2 0x10a5759 in op_replace /home/oni-link/git/neovim/src/nvim/ops.c:1696:14
    #3 0xfa9924 in do_pending_operator /home/oni-link/git/neovim/src/nvim/normal.c:1965:9
    #4 0x10694f3 in normal_finish_command /home/oni-link/git/neovim/src/nvim/normal.c:890:3
    #5 0xfdfd6c in normal_execute /home/oni-link/git/neovim/src/nvim/normal.c:1147:3
    #6 0x166f3d0 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:56:26
    #7 0xf94260 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #8 0xd55611 in main /home/oni-link/git/neovim/src/nvim/main.c:541:3
    #9 0x7fb5483095af in __libc_start_main (/lib64/libc.so.6+0x205af)

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:227 in __interceptor_strlen.part.25
```

File: src/nvim/ops.c [hunk 8]
```diff
@@ -1613,7 +1613,7 @@ int op_replace(oparg_T *oap, int c)
   int n, numc;
   int num_chars;
   char_u              *newp, *oldp;
-  size_t oldlen;
+  int oldlen;
```
Comment (oni-link) [inline src/nvim/ops.c]: `colnr_T` instead of `int` in case max line length is increased in the future.

**Comment by cacplate** [comment]
Seems like it works. Thank you @oni-link for your help. I will rebase now.

**Comment by justinmk** [comment]
Thanks @cacplate and @oni-link !

**Pull request merged**